libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Spec.hs

```diff
@@ -0,0 +1,1324 @@
+3) class Specify
+4) In RelSpec add (RelOp musthave canhave neverhave) to replace subset, superset, disjoint, can also express membership
+5) A better story about fields in constraints. Maybe add FieldSpec type.
+-}
+
+-- =========================================================
+
+-- | used when we run tests, and we have to pick some concrete Era
+type TT = MaryEra C_Crypto
+
+seps :: [String] -> String
+seps xs = List.intercalate " " xs
+
+sepsP :: [String] -> String
+sepsP xs = "(" ++ List.intercalate " " xs ++ ")"
+
+sepn :: [String] -> String
+sepn xs = List.intercalate "\n   " xs
+
+-- ============================================================
+-- Operators for Size (Defined in TypeRep.hs)
+
+-- | Used in tests so things don't get too large
+atleastdelta :: Int
+atleastdelta = 5
+
+-- | Used in tests so things don't get too large
+--   If we can't find an era using things of size 10
+--   using things of size 100, isn't going to help.
+atmostany :: Int
+atmostany = 10
+
+maxSize :: Size -> Int
+maxSize SzAny = atmostany
+maxSize (SzLeast i) = i + atleastdelta
+maxSize (SzMost n) = n
+maxSize (SzRng _ j) = j
+maxSize (SzNever xs) = errorMess "SzNever in maxSize" xs
+
+minSize :: Size -> Int
+minSize SzAny = 0
+minSize (SzLeast n) = n
+minSize (SzMost _) = 0
+minSize (SzRng i _) = i
+minSize (SzNever xs) = errorMess "SzNever in minSize" xs
+
+genSize :: Gen Size
+genSize =
+  frequency
+    [ (1, SzLeast <$> pos)
+    , (1, SzMost <$> anyAdds)
+    , (4, (\x -> SzRng x x) <$> pos)
+    , (1, do lo <- anyAdds; hi <- greater lo; pure (SzRng lo hi))
+    ]
+
+-- | Only use this only where you know it is NOT SzNever
+genFromSize :: Size -> Gen Int
+genFromSize (SzNever _) = error "Bad call to (genFromSize(SzNever ..))."
+genFromSize SzAny = chooseInt (0, atmostany)
+genFromSize (SzRng i j) = chooseInt (i, j)
+genFromSize (SzLeast i) = chooseInt (i, i + atleastdelta)
+genFromSize (SzMost i) = chooseInt (0, i)
+
+testSoundSize :: Gen Bool
+testSoundSize = do
+  spec <- genSize
+  ans <- genFromSize spec
+  pure $ runSize ans spec
+
+testMergeSize :: Gen Bool
+testMergeSize = do
+  spec1 <- genSize
+  spec2 <- genSize
+  case (spec1 <> spec2) of
+    SzNever _xs -> pure True -- trace (unlines _xs) $ pure True
+    SzAny -> trace "Aways True RelSpec" pure True
+    spec -> do
+      ans <- genFromSize spec
+      pure $ runSize ans spec && runSize ans spec1 && runSize ans spec2
+
+-- =====================================================
+-- RelSpec
+
+-- | Stores information to describe sets with relational constraints such as
+--   equality (x == y) subset (x ⊆ y) disjointness ( x ∩ y = ∅) superset (x ⊇ y).
+--   The information stored is used to compute the sets. Some times it is about
+--   what is in the set, and sometimes it is about what is NOT in the set.
+data RelSpec era dom where
+  -- \^ There is no restriction on the domain. Denotes the universe.
```

Comment:
I think the \^  was added by fourmolize. This has appeared a number of times. Is that possible?

Reply:
Yes, totally, cause that is invalid syntax and haddock would choke on it. So it is possible that fourmolu fixed it, by escaping the `^` symbol

(Note that it is invalid because there is no type constructor above it, which is what `-- ^` denotes)